src/AzureAppConfigurationImpl.ts

```diff
@@ -60,25 +105,141 @@ export class AzureAppConfigurationImpl extends Map<string, unknown> implements A
-                    keyValues.push([trimmedKey, value]);
+                loadedSettings.push(setting);
+            }
+        }
+        return loadedSettings;
+    }
+
+    /**
+     * Update etag of watched settings from loaded data. If a watched setting is not covered by any selector, a request will be sent to retrieve it.
+     */
+    async #updateWatchedKeyValuesEtag(existingSettings: ConfigurationSetting[]): Promise<void> {
+        if (!this.#refreshEnabled) {
+            return;
+        }
+
+        for (const sentinel of this.#sentinels) {
+            const matchedSetting = existingSettings.find(s => s.key === sentinel.key && s.label === sentinel.label);
+            if (matchedSetting) {
+                sentinel.etag = matchedSetting.etag;
+            } else {
+                // Send a request to retrieve key-value since it may be either not loaded or loaded with a different label or different casing
+                const { key, label } = sentinel;
+                const response = await this.#getConfigurationSettingWithTrace({ key, label });
+                if (response) {
+                    sentinel.etag = response.etag;
+                } else {
+                    sentinel.etag = undefined;
                 }
             }
         }
+    }
+
+    async #loadSelectedAndWatchedKeyValues() {
+        const keyValues: [key: string, value: unknown][] = [];
+
+        const loadedSettings = await this.#loadSelectedKeyValues();
+        await this.#updateWatchedKeyValuesEtag(loadedSettings);
+
+        // process key-values, watched settings have higher priority
+        for (const setting of loadedSettings) {
+            const [key, value] = await this.#processKeyValues(setting);
+            keyValues.push([key, value]);
+        }
+
+        this.clear(); // clear existing key-values in case of configuration setting deletion
         for (const [k, v] of keyValues) {
             this.set(k, v);
         }
     }
 
+    /**
+     * Load the configuration store for the first time.
+     */
+    async load() {
+        await this.#loadSelectedAndWatchedKeyValues();
+
+        // Mark all settings have loaded at startup.
+        this.#isInitialLoadCompleted = true;
+    }
+
+    /**
+     * Refresh the configuration store.
+     */
+    public async refresh(): Promise<void> {
+        if (!this.#refreshEnabled) {
+            throw new Error("Refresh is not enabled.");
+        }
+
+        // if still within refresh interval/backoff, return
+        if (!this.#refreshTimer.canRefresh()) {
+            return Promise.resolve();
+        }
+
+        // try refresh if any of watched settings is changed.
+        let needRefresh = false;
+        for (const sentinel of this.#sentinels.values()) {
+            const response = await this.#getConfigurationSettingWithTrace(sentinel, {
+                onlyIfChanged: true
+            });
+
+            if (response === undefined || response.statusCode === 200) {

```

Comment:
What if the sentinel was initially not loaded, and its still not present on the server?
response will be undefined, but we should not trigger refresh because nothing has changed.

Reply:
that's exactly a case not covered, as in the comment says in the next line.  
- changed [covered]
- deleted [covered]
- created [not covered]

We should cover it. And my idea is:
Compare sentinel.etag with response?.etag, if not equal, then need refresh.

Cases should all be covered:
- changed: etag are different.
- deleted: sentinel.etag has value but response is undefined
- created: sentinel.etag is undefined, but response.etag has value

WDYT?

Reply:
Actually I think created case is covered. The case not being covered is when a sentinel key is "Not Created".
We only need to check the etag when response is undefined. 

changed: etag are different, so response will be 200 (already covered)
deleted: sentinel.etag has value but response is undefined (already covered)
created: sentinel.etag is undefined, but response.etag has value, so response will be 200 (already covered)
NOT created: sentinel.etag is undefined, and response is also undefined -> Not covered

So you can just check if sentinel.etag is undefined when response is undefined -> this condition should not trigger a refresh.

Reply:
fixed in 8acf87b, according to the idea above.

Reply:
yes, as you mentioned, "created" WAS already covered. The commit fixes the case "not created".

Reply:
> just check if sentinel.etag is undefined when response is undefined

It's actually identical to checking `sentinel.etag === response?.etag`. Although they should never equal if they both have values (otherwise 304 and not go into this block).